### PR TITLE
Support for fallback web3 providers; Fix resource leaks on Eth1 monitor restarts

### PR DESCRIPTION
**Pull request by zah**

### CODE REVIEW

File: beacon_chain/eth1/eth1_monitor.nim [hunk 1]
```diff
@@ -1150,12 +1155,13 @@ when hasGenesisDetection:
...
              preset: RuntimePreset,
-             web3Url: string,
+             web3Urls: seq[string],
              depositContractAddress: Eth1Address,
              depositContractDeployedAt: BlockHashOrNumber,
              eth1Network: Option[Eth1Network]): Future[Result[T, string]] {.async.} =
     try:
-      let dataProviderRes = await Web3DataProvider.new(depositContractAddress, web3Url)
+      var urlIdx = 0
+      let dataProviderRes = await Web3DataProvider.new(depositContractAddress, web3Urls[urlIdx])
```
Comment (arnetheduck) [inline beacon_chain/eth1/eth1_monitor.nim:1167]: raise something better if weburls are empty?

Comment (zah) [reply beacon_chain/eth1/eth1_monitor.nim:1167]: That's really a pre-condition for creating an `Eht1Monitor` and I've converted it to an assert.

**Comment by arnetheduck** [comment]
uh, would be nice to document the intended policy for eth1 rotation here - it's a bit hard to understand from the code but I'm _guessing_ that it sticks with one eth1 provider until it fails and then moves on to the next, and sticks with _that_ until it fails etc? or does it try each provider for each request?

**Comment by zah** [comment]
This can happen in the book in parallel of the release.

Indeed, the code will try to use a single provider at a time until it has been demonstrated that the provider is not working (this happens after several retries with exponential backoff). It's currently not intended as a load balancing feature. The code will have to be more complicated otherwise, because it will have to deal with one difficult situation where one of the providers is not as completely synced as the other.

**Comment by arnetheduck** [comment]
> This can happen in the book in parallel of the release.

it's not even for users, but just for devs to understand the flow - there's no comments there what the intended behavior is and it's _hard_ to follow/understand from the code alone because of said retries, backoffs, timeouts etc - it would just be nice if it was written down what it should be doing

**Comment by arnetheduck** [comment]
as far as behaviours, it's an open question whether it should "jump back" to the "first" provider at any point - one could argue for both - ie if it goes offline for a few minutes and comes back online, what should happen?

**Pull request merged**